enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/IdReusabilityCondition.java

```diff
@@ -0,0 +1,113 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.state.machines.id;
+
+import java.util.function.BooleanSupplier;
+
+import org.neo4j.causalclustering.core.consensus.LeaderLocator;
+import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.kernel.impl.transaction.log.TransactionIdStore;
+import org.neo4j.kernel.impl.util.Dependencies;
+import org.neo4j.kernel.impl.util.Listener;
+
+/**
+ * Determines whether it is safe to reuse freed ids, based on current leader and tracking its own transactions.
+ * This should guarantee that a single freed id only ends up on a single core.
+ */
+public class IdReusabilityCondition implements BooleanSupplier, Listener<MemberId>
+{
+    private static final BooleanSupplier ALWAYS_FALSE = () -> false;
+
+    private TransactionIdStore transactionIdStore;
+    private final Dependencies dependencies;
+    private final MemberId myself;
+
+    private volatile BooleanSupplier currentSupplier = ALWAYS_FALSE;
+
+    public IdReusabilityCondition( Dependencies dependencies, LeaderLocator leaderLocator, MemberId myself )
+    {
+        this.dependencies = dependencies;
+        this.myself = myself;
+        leaderLocator.registerListener( this );
+    }
+
+    @Override
+    public boolean getAsBoolean()
+    {
+        return currentSupplier.getAsBoolean();
+    }
+
+    @Override
+    public void receive( MemberId newLeader )
+    {
+        if ( myself.equals( newLeader ) )
+        {
+            // We just became leader
+            resolveTransactionIdStore();
+            currentSupplier = new LeaderIdReusabilityCondition( transactionIdStore );
+        }
+        else
+        {
+            // We are not the leader
+            currentSupplier = ALWAYS_FALSE;
+        }
+    }
+
+    private void resolveTransactionIdStore()
+    {
+        if ( transactionIdStore == null )
+        {
+            transactionIdStore = dependencies.provideDependency( TransactionIdStore.class ).get();
+        }
+    }
+
+    private static class LeaderIdReusabilityCondition implements BooleanSupplier
+    {
+        private final TransactionIdStore transactionIdStore;
+        private final long idHighestWhenBecameLeader;
+
+        private volatile boolean oldTransactionsApplied;
+
+        LeaderIdReusabilityCondition( TransactionIdStore transactionIdStore )
+        {
+            this.transactionIdStore = transactionIdStore;
+
+            // Get highest transaction id seen
+            this.idHighestWhenBecameLeader = transactionIdStore.getLastCommittedTransactionId();
```

Comment:
Since the transactions are in the consensus log and the new leader uses them to prove it has the longest log, the highest seen transactions IDs should correlate with a leader switch. If there is any contention about the leadership the transactions belonging to older terms will not be accepted since only leaders are allowed to append new transactions. 

I don't see how

> thinking it is the leader

is enough to get new transactions committed in the consensus log or how you will accept new transaction from another leader during this uncertain state.